### PR TITLE
Adds `Relation.Nullary.Recomputable` plus consequences

### LINKED ISSUE #2199: `⊥` is `Recomputable`
The current definition of `Data.Empty`, and its derivative sub-module `Data.Empty.Irrelevant` are the basis of the current (re-)definition of `Relation.Nullary.Negation` and ... everything else involving `⊥` and negation `¬ A` etc. 

What's at present missing, and responsible for the artificial distinction between 
* `Data.Empty.Irrelevant.⊥-elim : ∀ {w} {Whatever : Set w} → .⊥ → Whatever`
* `Data.Empty.⊥-elim : ∀ {w} {Whatever : Set w} → ⊥ → Whatever`
 
is the equivalence of `.⊥` and `⊥`... ie that `⊥` is `Recomputable`.

NB `Recomputable` seems to be the modern-dress version of the very old idea that certain types (negated formulae, among others) have no computational content, that is, we can't make essential use (eg by pattern-matching) of their inhabitants in order to define/prove... other things. 

Well, the use of pattern-matching against the empty pattern `()` ensures that `⊥` itself is `Recomputable`:
```agda
Recomputable : ∀ {a} (A : Set a) → Set a
Recomputable A = .A → A

-- ⊥ is Recomputable

⊥-recompute : Recomputable ⊥
⊥-recompute ()
```
... and this is already provable in the current formulation of the library... and hence that the above two elimination principles are interderivable (so: do we need both, or does one suffice? UPDATED: does the asymmetry in the use of irrelevant marks in types between assumptions and conclusions mean that this cannot precisely be the case?). See [branch](https://github.com/jamesmckinna/agda-stdlib/tree/recomputable-empty), with no (seemingly!?) wider knock-on consequences for the rest of the library. 

So: we should reconcile these things, and their many consequences; esp. wrt proving `Recomputable` for all kinds of types, notably product types (below), negated types, decidable types, ... in general at least (conjecture!?) the (hereditarily) *Harrop formulas*, and more generally, any 'Glivenko class' of formulas (ditto.). 

Question/Issue: Why haven't we done this before/already? cf. #645 / #652 / #762

### PR DESCRIPTION
**Pull request by jamesmckinna**
Fixes #2199 , and the `Recomputable` and `Irrelevant` instances of #2091 .
Cherry-picks some of the lower-hanging fruit from #2149 . 

Outstanding issues:
- [x] names of things: `recompute` vs. `recomputable` vs. `...⇒recomputable` in all the various places; also the various additions to `Relation.Nullary.Negation.Core`; UPDATED sticking with `recompute` for now
- [x] deprecations: module `Data.Empty.Irrelevant` could now be deprecated entirely, but I have not (yet) done so; similarly, there may be knock-on deprecations arising from name choices above; UPDATED no deprecations!
- [x] where things should live/be (re-)exported from. UPDATED see @JacquesCarette 's comments and my revised commits

(Definite~~Probab~~ly) Won't do in relation to issues raised as a result of #2199 :
* "induction on the negative fragment": a separate (relativised!) `Data.Universe` construction for the 'Harrop' universe, as `Relation.Nullary.Recomputable.Universe`, with a view to generalising Harrop's results to ones of the form "if everything in `𝕌 = (U, El)` is `Recomputable`/`Irrelevant`/`Decidable`... then so too is everything in `Harrop 𝕌` etc. (UPDATED: downstream PR, but only once I've figured it out)
* relationship with the 'Irrelevance Axiom' or @nad 's developments mentioned there. (Ditto.)

### CODE REVIEW

File: src/Relation/Nullary/Decidable/Core.agda
Comment (jamesmckinna) [inline src/Relation/Nullary/Decidable/Core.agda:72]: Should this be renamed to `dec⇒recomputable` on the model of `Relation.Unary.Consequences`?

File: src/Relation/Nullary/Reflects.agda
Comment (jamesmckinna) [inline src/Relation/Nullary/Reflects.agda:62]: Should this be renamed to `reflects⇒recomputable` on the model of `Relation.Unary.Consequences` and `dec⇒recomputable`  in `Relation.Nullary.Decidable.Core` above?

Comment (JacquesCarette) [reply src/Relation/Nullary/Reflects.agda:62]: I must admit that I rather like 'recompute' for this one. But it's not a strong opinion.

File: src/Relation/Nullary/Decidable/Core.agda
Comment (gallais) [reply src/Relation/Nullary/Decidable/Core.agda:72]: Depends whether we see `Dec` as a predicate or a datatype
e.g. we use the name `toVec` rather than `list⇒vec`.

File: src/Relation/Nullary/Recomputable.agda
Comment (gallais) [inline src/Relation/Nullary/Recomputable.agda:39]: Can we add a proof that sigma is *not* recomputable?
Or is that a meta-theoretical result?

File: src/Relation/Nullary/Decidable/Core.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Decidable/Core.agda:72]: Nice analysis! (as with our other discussion on #2059 regarding `toSum`/`toDec` etc.)

I guess this means that, here at least, I have an inflection towards 'property' over 'datatype' (hence also my raising of #2091 )... but whether that's an 'innate' choice, or conditioned by/contingent on all of this happening under the aegis of `Nullary`, I'm not sure.

Cf. `Maybe` as a datatype constructor; `WeaklyDecidable` as its `Nullary` property incarnation...

File: src/Relation/Nullary/Recomputable.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Recomputable.agda:39]: Oh... great question (there's even actually something funny about the `_×_` case, in that it requires use of the projections, where I might have thought in the presence of eta that pattern-matching would suffice).

Not sure of the answer: the 'classical' picture was/is that what I've started calling the 'Harrop universe' excludes `∃` and `_⊎_`, towards Harrop's theorem about EP and DP being provable even under hypotheses in CL, provided all the assumptions are Harrop...

Danel Ahman remarked to me that all of this was defining a 'funny' universe of containers, so I suppose it's reasonable to ask what other container(functor)s enjoy the Harrop property, but that wasn't my original motivation, which was to understand the notion of being `Recomputable` as a counterpart to the orthodox realisability notion of 'computationally irrelevant' (as distinct from the mere `.` annotation, or more recent efforts on [Definitional Proof-Irrelevance Without K](https://inria.hal.science/hal-01859964v2/document)...).

Long term, what I want is a tactic which can take suitable (and easy to search for) _classical_ proofs of implications, and turn them into intuitionistically/type-theoretically acceptable ones (hence my question on the Zulip about Aczel/Smith-style slash interpretations, which do I think extend the fragment beyond Harrop, but ... It's All A Long Time Ago), by working with the Harrop metatheorem instantiated to the `Decidable`/`Stable` case.

But these things are (to me, at least) interesting research avenues, and (hence?) for the time being, out of scope for the PR ;-). But if you know/can say more about `Σ`, I'd be very interested!

**Comment by jamesmckinna** [comment]
Converting to DRAFT for the time being as I haven't had the brain space to revisit the questions above.

File: src/Relation/Nullary/Negation/Core.agda
Comment (jamesmckinna) [inline src/Relation/Nullary/Negation/Core.agda:70]: Can be derived from `weak-contradiction` and `¬-recompute`, now so omitted.

File: src/Relation/Nullary/Reflects.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Reflects.agda:62]: Sticking with `recompute` for now...

**Comment by jamesmckinna** [comment]
Incoming commits tidy this up, and then ready for review.

File: src/Relation/Nullary/Decidable/Core.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Decidable/Core.agda:73]: Sticking with `recompute` (for the time being)...

**Comment by jamesmckinna** [comment]
I further conjecture on #2329 that all uses of `⊥-elim` can be replaced by their irrelevant version `⊥-elim-irr`... so question for the maintainers, are we prepared to embrace such a (breaking?) change to the type of `⊥-elim`?

**Comment by MatthewDaggitt** [comment]
> so question for the maintainers, are we prepared to embrace such a (breaking?) change to the type of ⊥-elim?

Yes, but in v3.0! I suggest we make an issue with the 3.0 milestone?

**Comment by jamesmckinna** [comment]
@MatthewDaggitt fair point, but my belief (Modulo the merge conflict) is that everything here is v2.1 compatible so I won't pursue further changes this time around... ;-)

**Comment by jamesmckinna** [comment]
@JacquesCarette after the merge of #2329 could you review this one please?

File: CHANGELOG.md
Comment (jamesmckinna) [reply CHANGELOG.md:414]: Fair point... but once I found myself using `(flip contradiction)` more than 3 times in a single module, I decided I wanted/needed the abbreviation. 

Interestingly, even that form alone is not quite enough, especially when it interacts with additional implicit arguments and/or functions not being fully eta-expanded. But indeed, I'm happy to take the pushback...

File: src/Relation/Nullary/Negation.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Negation.agda:75]: Well, this is one where the implicits (I think) screw things up. I'll recheck, and revert if you're correct.

File: src/Relation/Nullary/Reflects.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Reflects.agda:17]: A case of 'flip-flopping' (sic) and failing to on a final design choice. Good catch!

File: src/Data/Empty.agda
Comment (jamesmckinna) [reply src/Data/Empty.agda:12]: Here's a third, which I think is at once less invasive, and more congruent with the spirit of the proposal, which would be to move this 'atomic' instance to `Recomputable` instead? I think that still works without creating cycles, but I'll investigate.

Comment (jamesmckinna) [reply src/Data/Empty.agda:12]: And yes, I agree with your 'conceptual' analysis (esp. given my recent comments on `decToMaybe` on #2330 ... I'd let my thinking on this PR get stale in the meantime and was trying to the minimum possible/necessary to push it over the line. Thanks for the nudge to reconsider in less haste...

File: src/Relation/Nullary/Recomputable.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Recomputable.agda:21]: Perhaps you're right, but I personally am not that fond of modules containing at most one definition, given that the whole thing is only 40 lines long (I think you are even more extreme than me about 'small modules' and dependency thrashing...).

File: src/Relation/Nullary/Negation.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Negation.agda:75]: This one is fine: reverted.

File: src/Relation/Nullary/Decidable.agda
Comment (jamesmckinna) [inline src/Relation/Nullary/Decidable.agda:55]: The implicits break putting just `(flip contradiction ¬a) here:
```agda
/home/james/RESEARCH/AGDA/Agda-stdlib-dev-v2.1/src/Relation/Nullary/Decidable.agda:55,75-96
Cannot instantiate the metavariable _C_128 to solution
(λ ()) (invert [¬a] x) ≡ x since it contains the variable x
which is not in scope of the metavariable
when checking that the expression flip contradiction ¬a has type
Function.Definitions.StrictlyInverseˡ _≡_ (λ ()) ¬a
```
and I'm stubborn enough to want to not see `⊥-elim` *anywhere* outside `Relation.Nullary.Negation.Core` or `Data.Empty`... (part of the point of this PR)

File: src/Data/Empty.agda
Comment (jamesmckinna) [inline src/Data/Empty.agda:40]: Does anyone have a better name for this?

File: src/Relation/Nullary/Recomputable.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Recomputable.agda:23]: So for now, I've left things as they are, plus moving the two other instances to here from `Data.Empty` and `Relation.Nullary.Negation.Core`, which may rejig the dependency graph more to your liking.

Comment (jamesmckinna) [reply src/Relation/Nullary/Recomputable.agda:23]: And at least now it could easily be broken up into `Base` and `Properties` if you insisted on it... but that feels like an easy downstream refactoring...

**Review by JacquesCarette — APPROVED** [review]

File: src/Data/Empty.agda
Comment (JacquesCarette) [reply src/Data/Empty.agda:40]: To be explicit: I do not.

File: src/Relation/Nullary.agda
Comment (MatthewDaggitt) [inline src/Relation/Nullary.agda:28]: Unless I'm missing something, this is a breaking change and therefore we probably shouldn't do it in v2.1?

File: src/Relation/Nullary/Recomputable.agda
Comment (MatthewDaggitt) [inline src/Relation/Nullary/Recomputable.agda:29]: Empty section heading?

File: src/Relation/Nullary/Recomputable.agda
Comment (MatthewDaggitt) [inline src/Relation/Nullary/Recomputable.agda:26]: Maybe a small comment (one or two sentences) just explaining what `Recomputable` means in terms of irrelevance? For those new-comers to the library who may not know what irrelevance is in much detail?

File: src/Relation/Nullary/Recomputable.agda
Comment (MatthewDaggitt) [inline src/Relation/Nullary/Recomputable.agda:32]: I feel that this section heading raises more questions than it answers for those not well versed in constructive mathematics. It's fine to mention Harrop constructions in comments, but maybe keep it out of section headings?

File: src/Relation/Nullary/Negation/Core.agda
Comment (MatthewDaggitt) [inline src/Relation/Nullary/Negation/Core.agda:51]: I think the name would be `weakContradiction`?

File: src/Relation/Nullary/Negation/Core.agda
Comment (MatthewDaggitt) [inline src/Relation/Nullary/Negation/Core.agda:58]: Are these refactorings really necessary? It's not like they rely on the contradiction being weak? I would advocate for keeping them using `contradiction` except where necessary? Not to mention it's just plain shorter.

File: src/Relation/Nullary/Reflects.agda
Comment (MatthewDaggitt) [inline src/Relation/Nullary/Reflects.agda:107]: Again I wouldn't change unless necessary?

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]

File: src/Relation/Nullary.agda
Comment (jamesmckinna) [reply src/Relation/Nullary.agda:28]: Don't think so. 
Previously, `recompute` was defined in `Decidable.Core`, and not at all in `Reflects`.
Now it's defined in `Reflects` for `Reflects A b` instances, then used to re-define `recompute` in `Decidable.Core`, so it's simulatantouesly been newly introduced in `Reflects` and then hidden from clashing with the 'old' definition in `Decidable.Core` in this PR. So I think that's *not* a breaking change? Maybe the `CHANGELOG` needs fixing, but I think it's OK...?

File: src/Relation/Nullary/Negation/Core.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Negation/Core.agda:51]: `contradiction-irr` is at least congruent with `⊥-elim-irr` so I'll make that change for the time being

File: src/Relation/Nullary/Negation/Core.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Negation/Core.agda:58]: Have reverted.

File: src/Relation/Nullary/Recomputable.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Recomputable.agda:26]: Done in incoming commit.

File: src/Relation/Nullary/Recomputable.agda
Comment (jamesmckinna) [reply src/Relation/Nullary/Recomputable.agda:29]: Have telescoped/collapsed into the next section...

File: src/Relation/Nullary.agda
Comment (jamesmckinna) [reply src/Relation/Nullary.agda:28]: Do let me know if the present `CHANGELOG` correctly captures the refactoring (or not!)

**Comment by jamesmckinna** [comment]
Now : proofs computed using `recompute` are *irrelevant*, so UIP follows immediately from `DecidableEquality`.
I've added these in as further `low-hanging-fruit` cherry-picked from #2149

**Comment by jamesmckinna** [comment]
@MatthewDaggitt are you happy with this not being a `breaking` change now?

**Comment by jamesmckinna** [comment]
As I understand things, there are (now!) two outstanding issues:
* is this PR `breaking` or not? I claim it is not, because specifically: in the top-level `stdlib` API, the `public` export of `recompute` (and now, additionally, `recompute-constant}`) from `Relation.Nullary` is still that from `Relation.Nullary.Decidable` (and hence from  `Relation.Nullary.Decidable.Core`), and *not* that of the newly introduced `Relation.Nullary.Reflects.recompute{-constant}`.
* does https://github.com/agda/agda/issues/7193 affect the 'safety' of this PR? From my (limited) understanding of @dolio 's reply to my question there, the answer is also that it does not.  

Hope this helps! (And removes any further obstruction to merging...)

**Review by JacquesCarette — APPROVED** [review]
After review of the latest stuff: it continues to look good and, this time, does look backwards compatible as well.

**Review by andreasabel — APPROVED** [review]
(Reviewing the irrelevance parts):  To me, these look like mainstream uses of irrelevance, not affected by agda/agda#7193

**Review by JacquesCarette — APPROVED** [review]

**Comment by JacquesCarette** [comment]
Just need @gallais to re-review.

**Comment by jamesmckinna** [comment]
> Just need ... to re-review.

And @MatthewDaggitt ?

**Review by MatthewDaggitt — APPROVED** [review]

**Pull request merged**
